File: exercise/php/day13/src/SantaMarket/Discount.php
<?php

namespace SantaMarket;

class Discount
{
    private string $description;
    private float $discountAmount;
    private Product $product;

    public function __construct(Product $product, string $description, float $discountAmount)
    {
        $this->product = $product;
        $this->description = $description;
        $this->discountAmount = $discountAmount;
    }

    public function equals(Discount $other, float $tolerance = 0.001): bool
    {
        return $this->product->equals($other->getProduct()) &&
            $this->description === $other->getDescription() &&
            $this->isCloseEnough($this->discountAmount, $other->getDiscountAmount(), $tolerance);
    }

    public function getProduct(): Product
    {
        return $this->product;
    }

    public function getDescription(): string
    {
        return $this->description;
    }

    private function isCloseEnough(float $a, float $b, float $tolerance): bool
    {
        return abs($a - $b) <= $tolerance;
    }

    public function getDiscountAmount(): float
    {
        return $this->discountAmount;
    }
}

File: solution/day13/src/main/java/santamarket/model/Discount.java
package santamarket.model;

import com.google.common.math.DoubleMath;
import java.util.Objects;

public class Discount {
    private final String description;
    private final double discountAmount;
    private final Product product;

    public Discount(Product product, String description, double discountAmount) {
        this.product = product;
        this.description = description;
        this.discountAmount = discountAmount;
    }

    public String getDescription() {
        return description;
    }

    public double getDiscountAmount() {
        return discountAmount;
    }

    public Product getProduct() {
        return product;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Discount discount = (Discount) o;
        return DoubleMath.fuzzyCompare(discount.discountAmount, discountAmount, 0.001) == 0 && Objects.equals(description, discount.description) && Objects.equals(product, discount.product);
    }

    @Override
    public int hashCode() {
        return Objects.hash(description, discountAmount, product);
    }

    @Override
    public String toString() {
        return "Discount{" +
                "description='" + description + '\'' +
                ", discountAmount=" + discountAmount +
                ", product=" + product +
                '}';
    }
}
